### PR TITLE
API Shift basic auth checking into middleware

### LINKED ISSUE #7554: Move BasicAuth::protect_site_if_necessary() into Middleware
Currently Controller is responsible for invoking this action. Similar to forceSSL / forceWWW, this should be handled via middleware.

```php
protected function init()
    {
        if ($this->basicAuthEnabled) {
            BasicAuth::protect_site_if_necessary();
        }

        // This is used to test that subordinate controllers are actually calling parent::init() - a common bug
        $this->baseInitCalled = true;
    }
```

# PRs

 - [x] https://github.com/silverstripe/silverstripe-framework/pull/7698

### PR DESCRIPTION
**Pull request by tractorcow**
Fixes #7554

Per-controller basicauth configuration is replaced with a route-matching rule applied to middleware.

### CODE REVIEW

**Comment by dhensby** [comment]
How do you force basic auth on a controller who's URL is dynamic (ie: a content controller)?

**Comment by tractorcow** [comment]
You manually put the basic auth call into your controller if you need to.

**Comment by dhensby** [comment]
I think it'd be better to match basic auth against Controllers rather than URL patterns. Or at least both. but this seems a bit of a step backwards if you can't protect a controller through the middleware...

File: src/Control/Controller.php [hunk 1]
```diff
@@ -528,9 +532,16 @@ public function render($params = null)
...
      * before it calls parent::init().
+     *
+     * @deprecated 4.0..5.0
      */
     public function disableBasicAuth()
     {
+        Deprecation::notice(
+            '5.0',
+            'Add this controller\'s url to SilverStripe\Security\BasicAuthMiddleware.ExcludeURLPatterns'
+            . ' property instead'
```
Comment (unclecheese) [inline src/Control/Controller.php]: Can we avoid the use of a hardcoded class? I'd prefer `BasicAuthMiddleware::class . 'ExcludeURLPatterns'`

File: src/Security/BasicAuthMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+     * @see CanonicalURLMiddleware
+     * @var array
+     */
+    protected $excludeURLPatterns = [];
+
+    /**
+     * URL to include, even if entire_site_protected is false.
+     * The list can either be an array of patterns to check, or an associative array
+     * of patterns mapped to either a permission code, or array of codes to check.
+     *
```
Comment (unclecheese) [inline src/Security/BasicAuthMiddleware.php]: This is a bit of a complex spec to describe with written docs only. As a dev using this for the first time, I'd much appreciate an example here, e.g.

```
[
  'my-secret-page' => 'ADMIN',
  '/^members-only\/*$',
]
```

File: _config/security.yml [hunk 3]
```diff
@@ -26,9 +26,14 @@ SilverStripe\Core\Injector\Injector:
...
       Middlewares:
         AuthenticationMiddleware: '%$SilverStripe\Security\AuthenticationMiddleware'
+        BasicAuthMiddleware: '%$SilverStripe\Security\BasicAuthMiddleware'
   SilverStripe\Security\AuthenticationMiddleware:
     properties:
       AuthenticationHandler: '%$SilverStripe\Security\AuthenticationHandler'
+  SilverStripe\Security\BasicAuthMiddleware:
+    properties:
+      ExcludeURLPatterns: []
+      IncludeURLPatterns: []
```
Comment (unclecheese) [inline _config/security.yml]: Just curious, what if I want my project to explicitly include a path that is excluded by a module, or vice-versa? Should we enforce an associative array here instead?


_security.yml_
```
SilverStripe\Security\BasicAuthMiddleware:
  properties:
    ExcludeURLPatterns:
        'my-path': true
        'my-other-path': 'ADMIN'
```

_mymodule.yml_
```
SilverStripe\Security\BasicAuthMiddleware:
  properties:
    ExcludeURLPatterns:
        'my-path': false
```

Comment (tractorcow) [reply _config/security.yml]: ok, so we use the bool flag for "enabled" right? And if it's a string, it's a permission check (except for exclude, which doesn't make sense to check for permissions).

Comment (tractorcow) [reply _config/security.yml]: Hm, it's a bit ambiguous if both key and value are strings; Someone might try to put the pattern as the value, and use a named key instead (like we do for extensions sometimes).

Comment (tractorcow) [reply _config/security.yml]: Ok, let's combine into one value; Keys are patterns, and value are:
 - bool true - force on (no permissions)
 - string - force on (special permission)
 - bool false - force exclude
 - null - default to global setting (controlled via yml)

Comment (tractorcow) [reply _config/security.yml]:
```php
/**
     * URL Patterns for basic auth. Keys are the Regexp string to match, and the key can
     * be one of the below:
     *  - true (bool) - Enabled for this url
     *  - false (bool) - Disabled for this url
     *  - Any string / array - Enabled for this url, and require the given string as a permission code
     *  - null (default) - Calls BasicAuth::protect_site_if_necessary(), which falls back to config setting
     *
     * @see CanonicalURLMiddleware
     * @var array
     */
    protected $urlPatterns = [];
```

File: src/Security/BasicAuthMiddleware.php (hunk 2 above)
Comment (tractorcow) [reply src/Security/BasicAuthMiddleware.php]: I've simplified this to a single array mapping patterns to rules.

**Comment by tractorcow** [comment]
> I think it'd be better to match basic auth against Controllers rather than URL patterns. Or at least both. but this seems a bit of a step backwards if you can't protect a controller through the middleware...

We can't do that in the middleware because there is no controller at this point. The controller is constructed, but it's not registered anywhere the middleware can access. Trying to make middleware handler-aware isn't really possible.

If you need this in your controller, just put it in your `MyController::init()` manually and skip middleware.

**Comment by tractorcow** [comment]
Updated based on feedback @unclecheese . I think the single-array config is a lot simpler and easier to understand.

File: _config/security.yml (hunk 3 above)
Comment (unclecheese) [reply _config/security.yml]: Brilliant. Love this change.

**Comment by dhensby** [comment]
> We can't do that in the middleware because there is no controller at this point. The controller is constructed, but it's not registered anywhere the middleware can access. Trying to make middleware handler-aware isn't really possible.
> 
> If you need this in your controller, just put it in your MyController::init() manually and skip middleware.

ok

**Comment by dhensby** [comment]
Behat tests are failing

File: src/Control/Controller.php [hunk 4]
```diff
@@ -61,6 +63,8 @@ class Controller extends RequestHandler implements TemplateGlobalProvider
     protected $templates = [];
 
     /**
+     * @deprecated 4.1.0...5.0.0 Add this controller's url to
+     * SilverStripe\Security\BasicAuthMiddleware.ExcludeURLPatterns property instead of setting false
```
Comment (flamerohr) [inline src/Control/Controller.php]: old property still suggested `urlPatterns` is the new one right?

File: src/Control/Controller.php [hunk 5]
```diff
@@ -528,9 +533,15 @@ public function render($params = null)
...
      * called before Controller::init(). That is, you must call it in your controller's init method
      * before it calls parent::init().
+     *
+     * @deprecated 4.0..5.0
      */
     public function disableBasicAuth()
     {
+        Deprecation::notice(
+            '5.0',
+            'Add this controller\'s url to ' . BasicAuthMiddleware::class . '.URLPatterns property instead'
```
Comment (flamerohr) [inline src/Control/Controller.php]: does casing matter? I see `urlPatterns` and `URLPatterns` used interchangeably, could we keep it to one case?

File: src/Security/BasicAuthMiddleware.php [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+     * @param HTTPRequest $request
+     * @return bool|string|array|null boolean value if enabled/disabled explicitly for this request,
+     * or null if should fall back to config value. Can also provide an explicit string / array of permission
+     * codes to require for this requset.
+     */
+    protected function checkMatchingURL(HTTPRequest $request)
+    {
+        // Null if no permissions enabled
+        $patterns = $this->getUrlPatterns();
```
Comment (flamerohr) [inline src/Security/BasicAuthMiddleware.php]: mixed casing for method names, try to keep to one :)

File: src/Control/Controller.php (hunk 4 above)
Comment (tractorcow) [reply src/Control/Controller.php]: oops sorry.

File: src/Control/Controller.php (hunk 5 above)
Comment (tractorcow) [reply src/Control/Controller.php]: `URLPatterns` is public setter `$middleware->URLPatterns`. `$urlPatterns` is the protected var. :)

File: src/Security/BasicAuthMiddleware.php (hunk 6 above)
Comment (tractorcow) [reply src/Security/BasicAuthMiddleware.php]: oops yes

**Comment by tractorcow** [comment]
Clarified PHPDocs and fixed case.

File: src/Control/Controller.php (hunk 5 above)
Comment (flamerohr) [reply src/Control/Controller.php]: oh, you're right :D nice one

**Comment by tractorcow** [comment]
Need to fix those behat tests sorry. =(

**Comment by tractorcow** [comment]
I missed a really obvious bug... you can't call BasicAuth::protect_site_if_necessary() when a controller isn't available. The very same restriction I just described to @dhensby ...

**Pull request merged**